### DISCUSSION TITLE
Ability to save replays as video

### DISCUSSION THREAD

**Discussion by kerrytazi** [ppy/osu#12986] [Ideas] [answered]
Sometimes it is not suitable or even possible to share replays of game with "osu replay files". So it may be cool feature to save reaplays as video like avi/mp4 formats with choosable framerate and resolution.
This way much easier for end-user to just click "save as video" instead of installing screencapture programs.
And sometimes on low-end PC it is not even possible to record replay.

**Comment by peppy** [comment]
On a low end PC where it is not possible to save as video, it will be just as impossible for osu! to do so. There’s no magic zero-overhead method to save video. It would need a full background render workflow which would likely take minutes to hours.

I disagree with implementing this at an osu! level when operating systems have this functionality built in already at negligible performance overheads (both windows and macOS).

**Reply by JakeOpsu2** [reply]
Pepsy

**Comment by mzhang28** [comment]
i've been [working on a program](https://github.com/iptq/osureplay) to do this for some time, although it's still undergoing massive rewrite (read: progress has stopped) to integrate with libosu. The idea would to have it be calculating the offset in non-realtime so you can still achieve nice fps with a low-end computer.

(example of emitted video: https://streamable.com/2g9o2)

**Comment by peppy** [comment]
The way lazer is written, it would not be hard to do something like this. I just think it's an unnecessary feature for a game to implement itself when external solutions already exist.

**Comment by Rekkonnect** [comment]
It's never bad to have an extra feature despite the alternative options. It's just not of high priority, I would prefer to see this being implemented once the rest is done.

**Comment by peppy** [comment]
Disagree. Feature creep is definitely a real thing.

**Comment by mzhang28** [comment]
third-party plugin api when??

**Comment by ayrod** [comment]
This feature is the #1 thing you should be working on.
With replays, you can only share with people who already play the game. With "export as some_video_format", you can have your players post their clips on their Discord. Then sit back and watch your playerbase grow.

Using screen recording software is not a solution. It makes videos of terrible quality because it has to work in real time. And even if it worked fine, it still wouldn't be a real solution. When you say "use a third party", you lose 95% of people. The feature has to be directly available in osu!.

**Comment by peppy** [comment]
We have plenty of priorities. Finishing the game and restoring all existing features is higher priority than a new (not-so-easy-to-implement) feature for us.

We'll keep this on the back-burner though. You're welcome to start on this one if you feel skilled enough!

**Comment by UboidStrata** [comment]
Well, I am not _that knowledgeable_ about video rendering but isn't the necessary process just creating output frames and combining them together at the desired fps with the desired resolution? As in something _say..._ an FFmpeg would do or something, since when I work on animations and render scenes using manim—math animation library from the 3b1b GitHub repository—I can wait all I want, but in the end, I get that nice 60fps 1080p or 1440p.

So I don't know if this helps or not; nevertheless, the way I see it is to output rendered frame (mix of background video/picture and circle patterns at that instant) and store to collect and further combine with other rendered frames. Again if this works similar to how my rendering goes, then all what low-end PC users have to do is wait more for their rendered replays.

Then again no pressure on this feature since some of us are capable of using external software or "third parties" to record these replays yet at the cost of reduced quality. Personally, if I knew more about how video rendering works and how to make such apps and whatnot I would've started assisting, but I unfortunately don't so yea.

**Comment by bdach** [comment]
Given that we've closed many issues on the basis that osu! is not a music player, I'd also argue that osu! is not a video recorder, either. External tools are the way to go.

**Comment by UboidStrata** [comment]
Actually, yes nevermind, that was slightly off the general purpose of the game, so implementing it to be part of osu! is a bit of an overkill.

Welp, recording with OBS is not that bad after all. I'll look into it though since this piqued my interest in replay-to-video rendering.

Nonetheless, thanks for the quick reply.

**Comment by ThatOneCalculator** [comment]
@bdach most of these external tools are windows-exclusive though.

**Reply by bdach** [reply]
OBS is not.

**Reply by ThatOneCalculator** [reply]
> OBS is not.

But replaying and recording in real time is far less convenient than just having an "export to video" button. I don't want to set up OBS and switch my layout every time I want to send a quick video to a friend.

**Reply by frenzibyte** [reply]
Despite your comment about your OBS experience (feels like you're not using it properly). It's not like we have the bandwidth and capability to support and maintain a video recorder within the game itself. It's *quite* easier said than actually done, so please don't expect this being added to the game anytime soon (if at all).

**Reply by peppy** [reply]
Please send your friend a replay instead.

**Comment by peppy** [comment, marked as answer]
I'm going to take the liberty of locking this discussion. No further discussion is going to be helpful here.

If someone offers an implementation of this (and it isn't 10k LoC) we may consider it. Otherwise you have a few options:

- Use an external tool (windows game bar, nvidia software, OBS)
- Share your gameplay using a replay instead

**Comment by Number-3434** [comment]
Just wanted to add that someone has created an [online tool](https://ordr.issou.best) that converts an osu replay file (`.osr`) to a video format
